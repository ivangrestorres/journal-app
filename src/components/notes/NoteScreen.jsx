import React from "react";
import { useRef } from "react";
import { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { activeNote, startDeleting } from "../../actions/notes";
import { useForm } from "../../hooks/useForm";
import { NotesAppBar } from "./NotesAppBar";

export const NoteScreen = () => {
    const dispatch = useDispatch();

    const { active: note } = useSelector((state) => state.notes);

    const [formValues, handleInputChange, reset] = useForm(note);
    const { title, body } = formValues;

    const activeId = useRef(note.id);

    useEffect(() => {
        if (note.id !== activeId.current) {
            reset(note);
            activeId.current = note.id;
        }
    }, [note, reset]);

    useEffect(() => {
        dispatch(activeNote(formValues.id, { ...formValues }));
    }, [formValues, dispatch]);

    const handleDelete = () => {
        dispatch(startDeleting(note.id));
    };

    return (
        <div className={"notes_main-content"}>
            <NotesAppBar />
            <div className={"notes_content"}>
                <input
                    type="text"
                    placeholder={"Some awesome title"}
                    className={"notes_title-input"}
                    name={"title"}
                    value={title}
                    onChange={handleInputChange}
                    autoComplete={"off"}
                />
                <textarea
                    placeholder={"What happened today"}
                    value={body}
                    name={"body"}
                    onChange={handleInputChange}
                    className={"notes_textarea"}
                ></textarea>
                {note.url && (
                    <div>
                        <img
                            src={note.url}
                            alt="imagen"
                            className={"notes_image"}
                        />
                    </div>
                )}
            </div>
            <button onClick={handleDelete} className={"btn btn-danger"}>
                Delete
            </button>
        </div>
    );
};
