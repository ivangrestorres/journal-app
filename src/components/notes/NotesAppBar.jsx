import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { startSaveNote, startUploading } from "../../actions/notes";

export const NotesAppBar = () => {
    const dispatch = useDispatch();
    const { active: note } = useSelector((state) => state.notes);

    const handleSaveNote = () => {
        dispatch(startSaveNote(note));
    };

    const handleFileChange = ({ target }) => {
        const file = target.files[0];
        if (file) {
            dispatch(startUploading(file));
        }
    };

    const handlePicture = () => {
        document.querySelector("#fileSelector").click();
    };

    return (
        <div className={"notes_appbar"}>
            <span>28 de agosto de 2020</span>
            <input
                id={"fileSelector"}
                onChange={handleFileChange}
                type="file"
                style={{ display: "none" }}
            />
            <div>
                <button onClick={handlePicture} className={"btn"}>
                    Picture
                </button>
                <button onClick={handleSaveNote} className={"btn"}>
                    Save
                </button>
            </div>
        </div>
    );
};
