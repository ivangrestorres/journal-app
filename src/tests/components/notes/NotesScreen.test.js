import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import thunk from "redux-thunk";
import configureStore from "redux-mock-store";
import "@testing-library/jest-dom";
import { NoteScreen } from "../../../components/notes/NoteScreen";
import { activeNote } from "../../../actions/notes";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

jest.mock("../../../actions/notes", () => ({
    activeNote: jest.fn(),
}));

const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null,
    },
    notes: {
        active: { id: 1234, title: "Hola", body: "Mundo", date: 0 },
        notes: [],
    },
};
let store = mockStore(initState);
store.dispatch = jest.fn();

describe("Pruebas <NotesScreen />", () => {
    beforeEach(() => {
        store = mockStore(initState);
        jest.clearAllMocks();
    });

    const wrapper = mount(
        <Provider store={store}>
            <NoteScreen></NoteScreen>
        </Provider>
    );
    test("debe de mostrar correctamente", () => {
        expect(wrapper).toMatchSnapshot();
    });

    test("debe de disparar el active note", () => {
        wrapper.find("input[name='title']").simulate("change", {
            target: {
                name: "title",
                value: "Hola de nuevo",
            },
        });

        expect(activeNote).toHaveBeenLastCalledWith(1234, {
            body: "Mundo",
            title: "Hola de nuevo",
            id: 1234,
            date: 0,
        });
    });
});
