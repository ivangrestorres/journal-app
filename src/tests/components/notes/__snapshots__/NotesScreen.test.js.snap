// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas <NotesScreen /> debe de mostrar correctamente 1`] = `
<div
  className="notes_main-content"
>
  <div
    className="notes_appbar"
  >
    <span>
      28 de agosto de 2020
    </span>
    <input
      id="fileSelector"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes_content"
  >
    <input
      autoComplete="off"
      className="notes_title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="Hola"
    />
    <textarea
      className="notes_textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happened today"
      value="Mundo"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
